File: 055/cpp/main.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int m1 = -1;
        int m2 = 0;
        int m = 0;
        while (true) {
            for (int i = m1 + 1; i < m2 + 1; i++) {
                m = std::max(m, i + nums[i]);
                if (m >= nums.size() - 1) {
                    return true;
                }
            }
            if (m == m2) {
                return false;
            }
            m1 = m2;
            m2 = m;
        }
        return false;
    }
};

File: 055/rust/main.rs
use std::cmp;

impl Solution {
    pub fn can_jump(nums: Vec<i32>) -> bool {
        let mut m1 = -1;
        let mut m2 = 0;
        let mut m = 0;
        loop {
            for i in m1 + 1..m2 + 1 {
                m = cmp::max(m, i + nums[i as usize]);
                if m >= nums.len() as i32 - 1 {
                    return true;
                }
            }
            if m == m2 {
                return false;
            }
            m1 = m2;
            m2 = m;
        }
        return false;
    }
}
